File: minipack-core/src/test/java/org/minipack/core/example/UseUnpooledAllocator.java
/*
 * Copyright 2024 the MiniPack contributors
 * SPDX-License-Identifier: Apache-2.0
 */
package org.minipack.core.example;

import org.junit.jupiter.api.Test;
import org.minipack.core.BufferAllocator;
import org.minipack.core.MessageReader;
import org.minipack.core.MessageWriter;

import java.io.IOException;
import java.nio.channels.ReadableByteChannel;
import java.nio.channels.WritableByteChannel;

import static org.assertj.core.api.Assertions.assertThat;

@SuppressWarnings("unused")
class UseUnpooledAllocator extends Example {
  // -8<- [start:snippet]
  class Example {
    final BufferAllocator allocator =
        BufferAllocator.ofUnpooled( options -> options //(1)
            .maxByteBufferCapacity(Integer.MAX_VALUE)  //(2)
            .maxCharBufferCapacity(Integer.MAX_VALUE)
        );

    void read(ReadableByteChannel channel) throws IOException {
      try (var reader = MessageReader.of(
          channel,
          options -> options.allocator(allocator)) //(3)
      ) { /* read some values */ }
    }

    void close() {
      allocator.close(); //(4)
    }
  }
  // -8<- [end:snippet]

  @Test
  void test() throws IOException {
    var example = new Example();
    example.read(inChannel);
    example.close();
  }
}

File: minipack-kotlin/src/test/kotlin/org/minipack/kotlin/example/UseUnpooledAllocator.kt
/*
 * Copyright 2024 the MiniPack contributors
 * SPDX-License-Identifier: Apache-2.0
 */
package org.minipack.kotlin.example

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.nio.channels.WritableByteChannel
import org.minipack.core.BufferAllocator
import org.minipack.kotlin.BufferAllocators
import org.minipack.kotlin.MessageReaders
import org.minipack.kotlin.MessageWriters
import java.nio.channels.ReadableByteChannel

@Suppress("UNUSED_ANONYMOUS_PARAMETER")
class UseUnpooledAllocator : Example() {
  // -8<- [start:snippet]
  class Example {
    val allocator: BufferAllocator = BufferAllocators.ofUnpooled( //(1)
      maxByteBufferCapacity = Integer.MAX_VALUE, //(2)
      maxCharBufferCapacity = Integer.MAX_VALUE
    )

    fun read(channel: ReadableByteChannel) {
      MessageReaders.of(channel, allocator = allocator).use { reader -> //(3)
        // read some values
      }
    }

    fun close() {
      allocator.close() //(4)
    }
  }
  // -8<- [end:snippet]

  @Test
  fun test() {
    val example = Example()
    example.read(inChannel)
    example.close()
  }
}
